tooling/scripts/ensure-python.sh

```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env bash
+
+# Special license: Take literally anything you want out of this file. I don't
+# care. Consider it WTFPL licensed if you like.
+# Basically there's a lot of suffering encoded here that I don't want you to
+# have to go through and you should feel free to use this to avoid some of
+# that suffering in advance.
+
+set -o errexit
+set -o nounset
+set -x
+
+# shellcheck source=tooling/scripts/common.sh
+source $(dirname "$0")/common.sh
+
+# This is to guard against multiple builds in parallel. The various installers will tend
+# to stomp all over each other if you do this and they haven't previously successfully
+# succeeded. We use a lock file to block progress so only one install runs at a time.
+# This script should be pretty fast once files are cached, so the lost of concurrency
+# is not a major problem.
+# This should be using the lockfile command, but that's not available on the
+# containerized travis and we can't install it without sudo.
+# Is is unclear if this is actually useful. I was seeing behaviour that suggested
+# concurrent runs of the installer, but I can't seem to find any evidence of this lock
+# ever not being acquired.
```

Comment:
Surely a single Travis run only installs a single version of Python, in which case parallelism is never a problem. Is this only an issue for local builds?